File: src/test/java/org/zstacks/zbus/rpc/RpcPerf.java
package org.zstacks.zbus.rpc;

import java.util.concurrent.atomic.AtomicLong;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.zstacks.zbus.client.Broker;
import org.zstacks.zbus.client.broker.SingleBroker;
import org.zstacks.zbus.client.broker.SingleBrokerConfig;
import org.zstacks.zbus.client.rpc.RpcProxy;
import org.zstacks.zbus.rpc.biz.Interface;
import org.zstacks.znet.Helper;

class Task extends Thread{
	private static final Logger log = LoggerFactory.getLogger(Task.class);
	Interface biz;
	int loopCount = 10000; 
	long startTime;
	AtomicLong counter;
	AtomicLong failCounter;
	@Override
	public void run() {  
		for(int i=0;i<loopCount;i++){ 
			try { 
				long count = counter.incrementAndGet();
				biz.getUserScore();
				
				if(count%1000 == 0){
					long end = System.currentTimeMillis();
					String qps = String.format("%.2f", count*1000.0/(end-startTime));
					log.info("QPS: {}, Failed/Total={}/{}({}%)",
							qps, failCounter.get(), counter.get(), 
							failCounter.get()*1.0/counter.get()*100);
				} 
			} catch (Exception e) { 
				failCounter.incrementAndGet();
				log.info(e.getMessage(), e);
				log.info("total failure {} of {} request", failCounter.get(), counter.get());
			}
		}
	}
}

public class RpcPerf {
	public static void main(String[] args) throws Exception { 
		final String brokerAddress = Helper.option(args, "-b", "127.0.0.1:15555");
		final int threadCount = Helper.option(args, "-c", 100);
		final int loopCount = Helper.option(args, "-loop", 10000);
		final String serviceName = Helper.option(args, "-service", "MyRpc");
		final int timeout = Helper.option(args, "-timeout", 10000);
		
		SingleBrokerConfig config = new SingleBrokerConfig(); 
		config.setBrokerAddress(brokerAddress);
		config.setMaxTotal(threadCount);
		config.setMaxIdle(threadCount);  
		
		final Broker broker = new SingleBroker(config);
		
		RpcProxy proxy = new RpcProxy(broker); 
		String url = String.format("mq=%s&&timeout=%d", serviceName, timeout);
		
		Interface biz = proxy.getService(Interface.class, url);
		
		AtomicLong counter = new AtomicLong(0);
		AtomicLong faileCounter = new AtomicLong(0);
		final long start = System.currentTimeMillis();
		Task[] tasks = new Task[threadCount];
		for(int i=0;i<tasks.length;i++){
			tasks[i] = new Task();
			tasks[i].biz = biz;
			tasks[i].loopCount = loopCount;
			tasks[i].startTime = start;
			tasks[i].counter = counter;
			tasks[i].failCounter = faileCounter;
		}
		
		for(Task task : tasks){
			task.start();
		} 
		for(Task task : tasks){
			task.join();
		}
		
		System.out.println("===done===");
		broker.close(); 
	}
}

File: API2/Python/examples/perf/rpc_perf.py
#encoding=utf8
import sys  
from threading import Thread, Lock
from time import time
from zbus import SingleBroker, Rpc 

sys.path.append('../../') 


class AtomicCounter:
    def __init__(self, init_value):
        self.counter = init_value
        self.lock = Lock()
        
    def increase_and_get(self):
        try:
            self.lock.acquire()
            self.counter = self.counter + 1
        finally:
            self.lock.release()
            return self.counter
    def get(self):
        return self.counter



import logging
class Task:
    log = logging.getLogger(__name__)
    def __init__(self): 
        self.rpc = None
        self.start_time = None 
        self.loop_count = None
        self.counter = None
        self.fail_counter = None
        
    def run(self):
        if self.rpc is None: 
            return
        for i in xrange(self.loop_count):
            try:
                total = self.counter.increase_and_get()
                fail = self.fail_counter.get()
                self.rpc.getUserScore()
                if (total+1)%1000 == 0: 
                    qps = '%.2f'%(total/(time()-self.start_time))
                    msg = "QPS: %s, Failed/Total=%s/%s(%.2f%s)"%(qps, fail, total, fail*100.0/total, '%')
                    self.log.info(msg)
            except Exception, e:
                self.log.info(e)
                self.fail_counter.increase_and_get()
   
def task_run(task):
    task.run()
    

start_time = time()
thread_count = 100
thread_loop_count = 100000000

broker = SingleBroker(host='127.0.0.1', port=15555, maxsize=thread_count)
rpc = Rpc(broker=broker, 
          mq='MyRpc', 
          module='Interface',
          encoding='utf8',
          timeout=10)

counter = AtomicCounter(0)
fail_counter = AtomicCounter(0) 
threads = []
for i in range(thread_count):
    task = Task()
    task.rpc = rpc 
    task.start_time = start_time
    task.counter = counter
    task.fail_counter = fail_counter
    task.loop_count = thread_loop_count
    
    thread = Thread(target=task_run, args=(task,))
    threads.append(thread)
    
for thread in threads:
    thread.start()
    
for thread in threads:
    thread.join()
    
    







